File: src/tsid_to_trajectory_bridge.cpp
#include <memory>
#include <vector>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"
#include "trajectory_msgs/msg/joint_trajectory.hpp"
#include "trajectory_msgs/msg/joint_trajectory_point.hpp"

class TSIDToTrajectoryBridge : public rclcpp::Node
{
public:
  TSIDToTrajectoryBridge() : Node("tsid_to_trajectory_bridge")
  {
    // Joint names in EXACT order from both TSID config and simulation controller
    // (Now they match after fixing controllers.yaml!)
    joint_names_ = {
      "right_elbow", "right_shoulder_roll", "right_shoulder_pitch",
      "left_elbow", "left_shoulder_roll", "left_shoulder_pitch", 
      "head_pitch", "head_yaw",
      "right_ankle_pitch", "right_knee", "right_hip_yaw", "right_hip_roll", "right_hip_pitch",
      "left_ankle_pitch", "left_knee", "left_hip_yaw", "left_hip_roll", "left_hip_pitch"
    };

    // Joint orders now match exactly - no mapping needed!
    RCLCPP_INFO(this->get_logger(), "Joint orders match exactly - direct 1:1 mapping");

    // Publishers and subscribers
    trajectory_pub_ = this->create_publisher<trajectory_msgs::msg::JointTrajectory>(
      "/position_controller/joint_trajectory", 10);

    tsid_sub_ = this->create_subscription<std_msgs::msg::Float64MultiArray>(
      "/robosoccer/joint_commands", 10,
      std::bind(&TSIDToTrajectoryBridge::tsidCommandCallback, this, std::placeholders::_1));

    RCLCPP_INFO(this->get_logger(), "TSID to Trajectory Bridge initialized!");
    RCLCPP_INFO(this->get_logger(), "Subscribing to: /robosoccer/joint_commands");
    RCLCPP_INFO(this->get_logger(), "Publishing to: /position_controller/joint_trajectory");
  }

private:

  void tsidCommandCallback(const std_msgs::msg::Float64MultiArray::SharedPtr msg)
  {
    if (msg->data.size() != joint_names_.size()) {
      RCLCPP_WARN(this->get_logger(), "Received %zu joint commands, expected %zu", 
                  msg->data.size(), joint_names_.size());
      return;
    }

    // Create JointTrajectory message
    trajectory_msgs::msg::JointTrajectory traj_msg;
    traj_msg.header.stamp = this->now();
    traj_msg.header.frame_id = "";
    traj_msg.joint_names = joint_names_;

    // Create trajectory point
    trajectory_msgs::msg::JointTrajectoryPoint point;
    point.positions.resize(joint_names_.size());
    point.velocities.clear(); // Let controller handle velocities
    point.accelerations.clear();
    point.effort.clear();
    point.time_from_start = rclcpp::Duration::from_nanoseconds(100000000); // 0.1 seconds

    // Direct 1:1 mapping since joint orders now match exactly!
    for (size_t i = 0; i < joint_names_.size(); ++i) {
      point.positions[i] = msg->data[i];
    }

    traj_msg.points.push_back(point);

    // Publish trajectory
    trajectory_pub_->publish(traj_msg);
  }

  rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr trajectory_pub_;
  rclcpp::Subscription<std_msgs::msg::Float64MultiArray>::SharedPtr tsid_sub_;
  
  std::vector<std::string> joint_names_;        // Joint names (same order for both TSID and simulation)
};

int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TSIDToTrajectoryBridge>());
  rclcpp::shutdown();
  return 0;
} 

File: robosoccer_control/tsid_to_trajectory_bridge.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64MultiArray
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint
from sensor_msgs.msg import JointState
import numpy as np

class TSIDToTrajectoryBridge(Node):
    def __init__(self):
        super().__init__('tsid_to_trajectory_bridge')
        self.declare_parameter('trajectory_duration', 0.05)
        self.declare_parameter('max_velocity', 10.0)
        self.declare_parameter('max_acceleration', 50.0)
        self.trajectory_duration = self.get_parameter('trajectory_duration').get_parameter_value().double_value
        self.max_velocity = self.get_parameter('max_velocity').get_parameter_value().double_value
        self.max_acceleration = self.get_parameter('max_acceleration').get_parameter_value().double_value
        self.use_sim_time = self.get_parameter_or('use_sim_time', rclpy.Parameter('use_sim_time', value=False)).value
        self.joint_names = [
            "right_elbow", "right_shoulder_roll", "right_shoulder_pitch",
            "left_elbow", "left_shoulder_roll", "left_shoulder_pitch", 
            "head_pitch", "head_yaw",
            "right_ankle_pitch", "right_knee", "right_hip_yaw", "right_hip_roll", "right_hip_pitch",
            "left_ankle_pitch", "left_knee", "left_hip_yaw", "left_hip_roll", "left_hip_pitch"
        ]
        self.current_positions = [0.0] * len(self.joint_names)
        self.current_velocities = [0.0] * len(self.joint_names)
        self.last_command_time = self.get_clock().now()
        self.last_joint_state_time = self.get_clock().now()
        self.get_logger().info("Joint orders match exactly - direct 1:1 mapping")
        self.get_logger().info(f"Trajectory duration: {self.trajectory_duration:.3f} s")
        self.trajectory_pub = self.create_publisher(JointTrajectory, '/position_controller/joint_trajectory', 10)
        self.tsid_sub = self.create_subscription(Float64MultiArray, '/robosoccer/joint_commands', 10, self.tsid_command_callback)
        self.joint_states_sub = self.create_subscription(JointState, '/joint_states', 10, self.joint_states_callback)
        self.status_pub = self.create_publisher(Float64MultiArray, '/bridge_status', 10)
        self.get_logger().info("Enhanced TSID to Trajectory Bridge initialized!")
        self.get_logger().info("Subscribing to: /robosoccer/joint_commands")
        self.get_logger().info("Subscribing to: /joint_states")
        self.get_logger().info("Publishing to: /position_controller/joint_trajectory")
        self.get_logger().info("Publishing status to: /bridge_status")

    def joint_states_callback(self, msg):
        if len(msg.position) >= len(self.joint_names) and len(msg.velocity) >= len(self.joint_names):
            for i in range(len(self.joint_names)):
                self.current_positions[i] = msg.position[i]
                self.current_velocities[i] = msg.velocity[i]
            self.last_joint_state_time = self.get_clock().now()

    def tsid_command_callback(self, msg):
        if len(msg.data) != len(self.joint_names):
            self.get_logger().warn(f"Received {len(msg.data)} joint commands, expected {len(self.joint_names)}")
            return
        now = self.get_clock().now()
        if (now - self.last_joint_state_time).nanoseconds / 1e9 > 0.1:
            self.get_logger().warn("No recent joint states received. Waiting for joint state feedback.")
            return
        traj_msg = JointTrajectory()
        traj_msg.header.stamp = now.to_msg()
        traj_msg.header.frame_id = ""
        traj_msg.joint_names = self.joint_names
        point = JointTrajectoryPoint()
        point.positions = [0.0] * len(self.joint_names)
        point.velocities = [0.0] * len(self.joint_names)
        point.accelerations = [0.0] * len(self.joint_names)
        point.effort = []
        point.time_from_start.sec = int(self.trajectory_duration)
        point.time_from_start.nanosec = int((self.trajectory_duration % 1) * 1e9)
        for i in range(len(self.joint_names)):
            target_position = msg.data[i]
            current_position = self.current_positions[i]
            current_velocity = self.current_velocities[i]
            point.positions[i] = target_position
            position_error = target_position - current_position
            desired_velocity = position_error / self.trajectory_duration
            desired_velocity = max(-self.max_velocity, min(self.max_velocity, desired_velocity))
            point.velocities[i] = desired_velocity
            velocity_error = desired_velocity - current_velocity
            desired_acceleration = velocity_error / self.trajectory_duration
            desired_acceleration = max(-self.max_acceleration, min(self.max_acceleration, desired_acceleration))
            point.accelerations[i] = desired_acceleration
        traj_msg.points = [point]
        self.trajectory_pub.publish(traj_msg)
        self.last_command_time = now
        self.publish_status()

    def publish_status(self):
        status_msg = Float64MultiArray()
        status_msg.data = [0.0] * 4
        now = self.get_clock().now()
        status_msg.data[0] = (now - self.last_command_time).nanoseconds / 1e9
        status_msg.data[1] = (now - self.last_joint_state_time).nanoseconds / 1e9
        status_msg.data[2] = self.trajectory_duration
        status_msg.data[3] = float(len(self.joint_names))
        self.status_pub.publish(status_msg)

def main(args=None):
    rclpy.init(args=args)
    node = TSIDToTrajectoryBridge()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main() 